firedrake/assemble.py

```diff
@@ -91,21 +92,104 @@ def assemble(expr, tensor=None, bcs=None, form_compiler_parameters=None,
 
 
+def _assemble_scalar(expr, scalar, assemble_now, **kwargs):
+    if len(expr.arguments()) != 0:
+        raise ValueError("Can't assemble a 0-form with arguments")
+    if scalar:
+        raise ValueError("Can't assemble 0-form into existing tensor")
+
+    scalar = _make_scalar()
+    _assemble_expr(expr, scalar, diagonal=False, assemble_now=assemble_now, **kwargs)
+    return scalar.data[0]
+
+
+def _assemble_vector(expr, vector, *, diagonal, assemble_now, **kwargs):
+    if diagonal:
+        test, trial = expr.arguments()
+        if test.function_space() != trial.function_space():
+            raise ValueError("Can only assemble diagonal of 2-form if functionspaces match")
+    else:
+        test, = expr.arguments()
+    if vector:
+        if test.function_space() != vector.function_space():
+            raise ValueError("Form's argument does not match provided result tensor")
+        vector.dat.zero()
+    else:
+        vector = _make_vector(test)
+    _assemble_expr(expr, vector, diagonal=diagonal, assemble_now=assemble_now, **kwargs)
+    return vector
+
+
+def _assemble_matrix(expr, matrix, *, mat_type, sub_mat_type, bcs, options_prefix, appctx, form_compiler_parameters, assemble_now):
+    if matrix:
+        if mat_type != "matfree":
+            matrix.M.zero()
+        if matrix.a.arguments() != expr.arguments():
+            raise ValueError("Form's arguments do not match provided result tensor")
+    else:
+        matrix = _make_matrix(expr, mat_type=mat_type, sub_mat_type=sub_mat_type, bcs=bcs, options_prefix=options_prefix, appctx=appctx, form_compiler_parameters=form_compiler_parameters)
+
+    if mat_type == "matfree":
+        matrix.assemble()
+    else:
+        _assemble_expr(expr, matrix, diagonal=False, mat_type=mat_type, sub_mat_type=sub_mat_type, bcs=bcs, options_prefix=options_prefix, appctx=appctx, form_compiler_parameters=form_compiler_parameters, assemble_now=assemble_now)
+        matrix.M.assemble()
+    return matrix
+
+
+def _make_scalar():
+    return op2.Global(1, [0.0], dtype=utils.ScalarType)
+
+
+def _make_vector(test):
+    return firedrake.Function(test.function_space())
+
+
+def _assemble_expr(expr, tensor, *, bcs, diagonal, assemble_now, **kwargs):
+    bcs = _preprocess_bcs(bcs)
+    # TODO: Reimplement this when PyOP2 functionality added.
+    # if "parloops" in expr._cache:
+    #     parloops = expr._cache["parloops"]
+    # else:
+    #     parloops = _make_parloops(expr, tensor, bcs=bcs, diagonal=diagonal, **kwargs)
+    #     expr._cache["parloops"] = parloops
+    # parloop.compute(out=tensor.dat)
+    parloops = _make_parloops(expr, tensor, bcs=bcs, diagonal=diagonal, **kwargs)
+    for parloop in parloops:
+        parloop.compute()
+
+    assembly_rank = _get_assembly_rank(expr, diagonal)
+
+    dir_bcs = tuple(bc for bc in bcs if isinstance(bc, DirichletBC))
+    _apply_dirichlet_bcs(tensor, dir_bcs, assembly_rank, diagonal=diagonal, assemble_now=assemble_now)
```

Comment:
These also cons parloops, not sure if we want to save them too.

Reply:
We could cache the callables, but I think the parloops are stashed on the Dats so we're only saving a little bit.